### ISSUE TITLE
irrlicht:blocked on macOS

### ISSUE THREAD

**Issue by virtualritz** [microsoft/vcpkg#15011]
This port is blocked on macOS but this platform is supported. From [the Irrlicht website](http://irrlicht.sourceforge.net/?page_id=45):
```
The Irrlicht Engine is platform independent, currently there is official support for:
XP, XP64, Vista, CE, Windows 7,Windows 8 and Windows 10
Linux
OSX
Sun Solaris/SPARC
All platforms using SDL
```

SDL is also available on ARM so Irrlicht being blocked for that target is likely also a mistake but a separate issue.

**Comment by NancyLi1013** [comment]
Hi @virtualritz 

Thanks for posting this issue.

Currently, `irrlicht ` failed on osx due to these errors:

```
In file included from /Users/vcpkg/Nancy/vcpkg/buildtrees/irrlicht/src/de69ddd2c6-1cdb4cb3bd/source/Irrlicht/COpenGLDriver.cpp:21:
3391 /Users/vcpkg/Nancy/vcpkg/buildtrees/irrlicht/src/de69ddd2c6-1cdb4cb3bd/source/Irrlicht/MacOSX/CIrrDeviceMacOSX.h:12:1: error: unknown type name '_'
3392 _
3393 ^
3394 In file included from /Users/vcpkg/Nancy/vcpkg/buildtrees/irrlicht/src/de69ddd2c6-1cdb4cb3bd/source/Irrlicht/COpenGLDriver.cpp:21:
3395 In file included from /Users/vcpkg/Nancy/vcpkg/buildtrees/irrlicht/src/de69ddd2c6-1cdb4cb3bd/source/Irrlicht/MacOSX/CIrrDeviceMacOSX.h:13:
3396 In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSWindow.h:8:
3397 In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSArray.h:5:
3398 In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:7:
3399 /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:492:1: error: expected u     nqualified-id
3400 @class NSString, Protocol;
3401 ^
3402 /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:494:9: error: unknown ty     pe name 'NSString'
3403 typedef NSString * NSExceptionName NS_EXTENSIBLE_STRING_ENUM;
3404         ^
3405 /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:495:9: error: unknown ty     pe name 'NSString'
3406 typedef NSString * NSRunLoopMode NS_EXTENSIBLE_STRING_ENUM;
```
But I don't know how to fix this problem.

@strega-nil  Do you have any suggestions about this issue?

**Comment by strega-nil** [comment]
This I do not know. I can look at it next week, perhaps?

**Comment by shauser67** [comment]
It looks like you are trying to compile Irrlicht/COpenGLDriver.cpp with a C++ compiler, but you have to use an ObjC compiler instead:

[Getting error while using ObjC class include in .cpp file (unqualified-id @class)](https://stackoverflow.com/questions/46001065/getting-error-while-using-objc-class-include-in-cpp-file-unqualified-id-class)
[
iOS build error: expected unqualified-id @class NSString](https://github.com/HaxeFoundation/hxcpp/issues/615)

**Comment by traversaro** [comment]
> 
> 
> Hi @virtualritz
> 
> Thanks for posting this issue.
> 
> Currently, `irrlicht ` failed on osx due to these errors:
> 
> ```
> In file included from /Users/vcpkg/Nancy/vcpkg/buildtrees/irrlicht/src/de69ddd2c6-1cdb4cb3bd/source/Irrlicht/COpenGLDriver.cpp:21:
> 3391 /Users/vcpkg/Nancy/vcpkg/buildtrees/irrlicht/src/de69ddd2c6-1cdb4cb3bd/source/Irrlicht/MacOSX/CIrrDeviceMacOSX.h:12:1: error: unknown type name '_'
> 3392 _
> 3393 ^
> 3394 In file included from /Users/vcpkg/Nancy/vcpkg/buildtrees/irrlicht/src/de69ddd2c6-1cdb4cb3bd/source/Irrlicht/COpenGLDriver.cpp:21:
> 3395 In file included from /Users/vcpkg/Nancy/vcpkg/buildtrees/irrlicht/src/de69ddd2c6-1cdb4cb3bd/source/Irrlicht/MacOSX/CIrrDeviceMacOSX.h:13:
> 3396 In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSWindow.h:8:
> 3397 In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSArray.h:5:
> 3398 In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:7:
> 3399 /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:492:1: error: expected u     nqualified-id
> 3400 @class NSString, Protocol;
> 3401 ^
> 3402 /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:494:9: error: unknown ty     pe name 'NSString'
> 3403 typedef NSString * NSExceptionName NS_EXTENSIBLE_STRING_ENUM;
> 3404         ^
> 3405 /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:495:9: error: unknown ty     pe name 'NSString'
> 3406 typedef NSString * NSRunLoopMode NS_EXTENSIBLE_STRING_ENUM;
> ```
> 
> But I don't know how to fix this problem.
> 
> @strega-nil Do you have any suggestions about this issue?

I do not know if it related, but the homebrew's Formula of irrlicht has two macOS-specific patches to the 1.8.4 codebase, one of which is related to the `source/Irrlicht/MacOSX/CIrrDeviceMacOSX.m` source code, perhaps it could make sense to try to use them also in the vcpkg port of irrlicht.

**Comment by traversaro** [comment]
Furthermore, the homebrew formula is using the included Xcode build system (https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/irrlicht.rb#L40), while the vcpkg port is using a custom CMake build system.

**Comment by traversaro** [comment]
Another useful reference is patch https://sourceforge.net/p/irrlicht/patches/320/ by @shybovycha that provides a CMake build script for irrlicht that seems to support macOS as well.

**Comment by traversaro** [comment]
I am not currently interested in having irrlicht working on vcpkg on macOS, but for anyone interested in getting this to work in https://github.com/conda-forge/staged-recipes/pull/13900 I added a conda-forge recipe for irrlicht, and the fix to the CMake build system to fix the compilation were the following.

* Remove the `-no-rtti` flag in https://github.com/microsoft/vcpkg/blob/2020.11/ports/irrlicht/CMakeLists.txt#L190 
* Add a block to add .mm files to the compilation and enable Objective C++ compilation for mixed files (https://github.com/conda-forge/staged-recipes/pull/13900/files#diff-2c220af85fe5d0a51fea98e69ead2630be651545ff3f2912005aa5cc91cfb6d7R99):
~~~cmake
if(APPLE)
    file(GLOB IRR_SRC_FILES_MACOS "${IRR_SRC_DIR}/MacOSX/*.mm" "${IRR_SRC_DIR}/MacOSX/*.h" )
    list(APPEND IRR_ALL_SRC_FILES ${IRR_SRC_FILES_MACOS})
    set_source_files_properties("${IRR_SRC_DIR}/Irrlicht.cpp" PROPERTIES COMPILE_OPTIONS "-x;objective-c++")
    set_source_files_properties("${IRR_SRC_DIR}/COpenGLDriver.cpp" PROPERTIES COMPILE_OPTIONS "-x;objective-c++")
endif()
~~~
* Add a block to link macOS's System Frameworks ( https://github.com/conda-forge/staged-recipes/pull/13900/files#diff-2c220af85fe5d0a51fea98e69ead2630be651545ff3f2912005aa5cc91cfb6d7R225 ):
~~~cmake
    if(APPLE)
        find_library(COCOA_LIB Cocoa)
        find_library(CARBON_LIB Carbon)
        find_library(IOKIT_LIB IOKit)
        target_link_libraries(Irrlicht PRIVATE ${IOKIT_LIB} ${COCOA_LIB} ${CARBON_LIB})
        target_compile_definitions(Irrlicht PRIVATE MACOSX)
        target_include_directories(Irrlicht PRIVATE ${IRR_SRC_DIR}/MacOSX)
    endif()
~~~

**Comment by NancyLi1013** [comment]
Hi all, 

`irrlicht `has supported osx in vcpkg now, please try to update to the latest vcpkg to rebuild this port.
